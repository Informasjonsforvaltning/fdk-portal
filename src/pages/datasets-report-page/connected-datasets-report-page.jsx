import { connect } from 'react-redux';
import { fetchPublishersIfNeededAction } from '../../redux/modules/publishers';
import { DatasetsReportPage } from './datasets-report-page';

const mapStateToProps = ({ publishers }) => ({
  publishers: publishers && publishers.publisherItems
});

const mapDispatchToProps = dispatch => ({
  fetchPublishersIfNeeded: () => dispatch(fetchPublishersIfNeededAction())
});

export const ConnectedDatasetsReportPage = connect(
  mapStateToProps,
  mapDispatchToProps
)(DatasetsReportPage);
