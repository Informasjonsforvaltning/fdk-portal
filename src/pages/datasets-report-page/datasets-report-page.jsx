import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import qs from 'qs';
import { Button } from 'reactstrap';

import localization from '../../lib/localization';
import { PublishersSelect } from './publishers-select/publishers-select.component';
import { PublishersTree } from './publishers-tree/publishers-tree.component';
import { getParamFromLocation } from '../../lib/addOrReplaceUrlParam';
import { ResolvedReportStats } from './report-stats/resolved-report-stats';
import { isFilterActive } from './filter-helper';

export function DatasetsReportPage({
  location,
  history,
  fetchPublishersIfNeeded,
  publishers
}) {
  function selectPublisher(publisher) {
    const orgPath = publisher && publisher.orgPath;
    const currentSearch = qs.parse(location.search, {
      ignoreQueryPrefix: true
    });
    const newSearch = { ...currentSearch, orgPath };
    const newSearchStr = qs.stringify(newSearch, {
      addQueryPrefix: true,
      skipNulls: true,
      encode: false
    });

    // This is react-router browserHistory object
    // https://github.com/ReactTraining/history
    history.push({ search: newSearchStr });
  }

  function clearSearch() {
    selectPublisher(null);
  }

  fetchPublishersIfNeeded();

  const orgPath = getParamFromLocation(location, 'orgPath');
  const selectedPublisher = _.get(publishers, [orgPath], null);

  return (
    <section className="container">
      <div className="row">
        <div className="col-md-4">
          {isFilterActive({ orgPath }) && (
            <Button
              className="fdk-button fade-in-500"
              onClick={clearSearch}
              color="primary"
            >
              {localization.query.clear}
            </Button>
          )}
          <PublishersSelect
            publishers={publishers}
            onChange={selectPublisher}
            value={selectedPublisher}
          />
          <PublishersTree
            onChange={selectPublisher}
            value={selectedPublisher}
          />
        </div>
        <div className="col-md-8">
          <ResolvedReportStats
            orgPath={selectedPublisher && selectedPublisher.orgPath}
            entityName={selectedPublisher && selectedPublisher.name}
          />
        </div>
      </div>
    </section>
  );
}

DatasetsReportPage.defaultProps = {
  fetchPublishersIfNeeded: _.noop,
  publishers: {}
};

DatasetsReportPage.propTypes = {
  fetchPublishersIfNeeded: PropTypes.func,
  location: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  publishers: PropTypes.object
};
