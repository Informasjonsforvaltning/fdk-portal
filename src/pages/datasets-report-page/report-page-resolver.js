import _ from 'lodash';
import { resolve } from 'react-resolver';
import { getDatasetStats } from '../../api/get-dataset-stats';
import { getApiStats } from '../../api/get-api-stats';
import { getConceptStats } from '../../api/get-concept-stats';
import { conceptsSearch, extractConcepts } from '../../api/concepts';
import { getParamFromLocation } from '../../lib/addOrReplaceUrlParam';

const memoizedGetDatasetStats = _.memoize(getDatasetStats);
const memoizedGetApiStats = _.memoize(getApiStats);
const memoizedGetConceptStats = _.memoize(getConceptStats);
const memoizedSearchConcepts = _.memoize(conceptsSearch, JSON.stringify);

const mapProps = {
  datasetStats: ({ location }) =>
    memoizedGetDatasetStats(getParamFromLocation(location, 'orgPath')),
  apiStats: ({ location }) =>
    memoizedGetApiStats(getParamFromLocation(location, 'orgPath')),
  conceptStats: ({ location }) =>
    memoizedGetConceptStats(getParamFromLocation(location, 'orgPath')),
  mostUsedConcepts: async ({ location }) => {
    const conceptStats = await memoizedGetConceptStats(
      getParamFromLocation(location, 'orgPath')
    );
    const datasetCountsByConceptUri = _.get(
      conceptStats,
      'datasetCountsByConceptUri',
      {}
    );

    const mostUsedConceptsArray = _.orderBy(
      datasetCountsByConceptUri,
      [conceptsWithDataset => conceptsWithDataset.length],
      ['desc']
    )
      .reduce((result, value) => {
        result.push(_.get(value, [0, 'subjectUri']));
        return result;
      }, [])
      .slice(0, 4);

    return Promise.resolve(
      memoizedSearchConcepts({ uris: mostUsedConceptsArray.join() })
    ).then(extractConcepts);
  }
};

export const reportPageResolver = resolve(mapProps);
