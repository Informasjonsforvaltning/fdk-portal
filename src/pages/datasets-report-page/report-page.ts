import { compose } from 'redux';

import { ReportPagePure } from './report-page-pure';
import { reportPageConnector } from './report-page-connector';
import { reportPageResolver } from './report-page-resolver';

const enhance = compose(reportPageConnector, reportPageResolver);

export const ReportPage = enhance(ReportPagePure);
