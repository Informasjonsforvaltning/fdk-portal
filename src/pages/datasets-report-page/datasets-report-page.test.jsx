import React from 'react';
import { shallow } from 'enzyme';
import { DatasetsReportPage } from './datasets-report-page';

test('should render DatasetsReportPage correctly', () => {
  const wrapper = shallow(
    <DatasetsReportPage
      location={{ search: '' }}
      history={{ push: () => {} }}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
