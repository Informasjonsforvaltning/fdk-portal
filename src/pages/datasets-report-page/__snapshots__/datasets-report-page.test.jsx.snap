// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DatasetsReportPage correctly 1`] = `
<section
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-md-4"
    >
      <PublishersSelect
        onChange={[Function]}
        publishers={Object {}}
        value={null}
      />
      <PublishersResolver
        onChange={[Function]}
        value={null}
      />
    </div>
    <div
      className="col-md-8"
    >
      <DatasetStatsApiStatsConceptStatsMostUsedConceptsResolver
        entityName={null}
        orgPath={null}
      />
    </div>
  </div>
</section>
`;
